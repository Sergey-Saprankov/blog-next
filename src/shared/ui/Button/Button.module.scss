.Button {
  cursor: pointer;

  line-height: 0;

  border: none;

  transition: all 0.2s ease-in-out;

  &:hover {
    transform: translateY(-1px);

    opacity: 0.5;
  }

  &:disabled {
    opacity: 0.3;
  }
}

.clear {
  margin: 0;
  padding: 0;

  color: var(--color-primary);

  background: none;
  border: none;
  outline: none;
}

.primary {
  color: var(--btn-primary-color);

  background: var(--btn-primary-bg);
}

.secondary {
  color: var(--btn-secondary-color);

  background: var(--btn-secondary-bg);
}

.destructive {
  color: var(--btn-primary-color);

  background: var(--btn-destructive-bg);
}

.large {
  width: var(--btn-size-width-large);
  height: var(--btn-size-height-large);

  border-radius: 24px;
}

.small {
  width: var(--btn-size-width-small);
  height: var(--btn-size-height-small);

  border-radius: 20px;
}
