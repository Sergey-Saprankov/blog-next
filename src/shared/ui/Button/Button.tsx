import React, { ButtonHTMLAttributes, FC, memo, ReactNode } from 'react'

import { classNames } from '../../lib/classNames/classNames'

import s from './Button.module.scss'

export enum ButtonTheme {
  Clear = 'clear',
  Primary = 'primary',
  Secondary = 'secondary',
  Destructive = 'destructive',
}

export enum ButtonSize {
  Small = 'small',
  Large = 'large',
}

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  className?: string
  children: ReactNode
  theme?: ButtonTheme
  size?: ButtonSize
}

export const Button: FC<ButtonProps> = memo(
  ({ className, children, theme = '', size = '', type = 'button', ...other }) => {
    return (
      <button
        type={type}
        {...other}
        className={classNames(s.Button, {}, [s[theme], s[size], className])}
      >
        {children}
      </button>
    )
  }
)
