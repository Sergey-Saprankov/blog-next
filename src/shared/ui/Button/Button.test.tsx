import { fireEvent, render, screen } from '@testing-library/react'

import { Button, ButtonSize, ButtonTheme } from './Button'

describe('Button component', () => {
  test('renders with default props', () => {
    render(<Button>Test</Button>)
    expect(screen.getByText('Test')).toBeInTheDocument()
  })

  test('renders with specified theme', () => {
    render(<Button theme={ButtonTheme.Secondary}>Test</Button>)
    expect(screen.getByText('Test')).toHaveClass('secondary')
  })

  test('theme, size test', () => {
    render(
      <Button size={ButtonSize.Small} theme={ButtonTheme.Clear}>
        Test
      </Button>
    )
    expect(screen.getByText('Test')).toHaveClass('small')
    screen.debug()
  })

  test('snapshot', () => {
    render(
      <Button size={ButtonSize.Small} theme={ButtonTheme.Clear}>
        Test
      </Button>
    )
    expect(screen.getByText('Test')).toMatchSnapshot()
    screen.debug()
  })

  test('calls onClick handler when clicked', () => {
    const handleClick = jest.fn()
    const { getByText } = render(<Button onClick={handleClick}>Click me</Button>)

    fireEvent.click(getByText('Click me'))
    expect(handleClick).toHaveBeenCalled()
  })
})
