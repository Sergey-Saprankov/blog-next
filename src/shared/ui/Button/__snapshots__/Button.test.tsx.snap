// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component snapshot 1`] = `
<button
  class="Button clear small"
>
  Test
</button>
`;

exports[`Button snapshot 1`] = `
<button
  class="Button clear small"
>
  Test
</button>
`;
